File: assets/rule-templates/consent-before-auto-action/python/fakes.py
"""Фейки реестра согласий, незавершённых операций и счётчика автодействий.

Фейки считают то, что проверяет инвариант: есть ли действующая запись
о согласии (`is_active`), открыта ли операция клиента (`has_open`) и сколько
автодействий создано (`actions`). Ни сети, ни сна — детерминированный
прогон за миллисекунды.
"""


class ConsentRegistry:
    """Реестр согласий: записи о согласии клиента на автодействие.

    Запись хранит статус, а не только факт наличия: отозванное согласие
    остаётся в реестре, но действующим не считается.
    """

    def __init__(self) -> None:
        self._records: dict[tuple[str, str], dict] = {}

    def grant(self, client_id: str, scope: str, consent_id: str) -> None:
        """Записать согласие клиента в объёме `scope`."""
        self._records[(client_id, scope)] = {"consent_id": consent_id, "status": "active"}

    def revoke(self, client_id: str, scope: str) -> None:
        """Отозвать согласие; запись остаётся со статусом `revoked`."""
        record = self._records.get((client_id, scope))
        if record is not None:
            record["status"] = "revoked"

    def is_active(self, client_id: str, scope: str) -> bool:
        """Есть ли действующая запись о согласии."""
        record = self._records.get((client_id, scope))
        return record is not None and record["status"] == "active"

    def record_for(self, client_id: str, scope: str):
        """Запись о согласии или None."""
        return self._records.get((client_id, scope))


class FakeOperations:
    """Незавершённые операции клиента.

    Пока операция клиента открыта, новое автодействие по нему не стартует.
    """

    def __init__(self) -> None:
        self._open: dict[str, str] = {}

    def open(self, client_id: str, operation_id: str) -> None:
        """Открыть операцию клиента."""
        self._open[client_id] = operation_id

    def close(self, client_id: str) -> None:
        """Закрыть операцию клиента."""
        self._open.pop(client_id, None)

    def has_open(self, client_id: str) -> bool:
        """Открыта ли у клиента незавершённая операция."""
        return client_id in self._open


class FakeActionLog:
    """Счётчик созданных автодействий."""

    def __init__(self) -> None:
        self.actions: list[tuple[str, str, int]] = []

    def create(self, client_id: str, kind: str, amount: int) -> dict:
        """Создать автодействие и записать его в счётчик."""
        self.actions.append((client_id, kind, amount))
        return {
            "client_id": client_id,
            "kind": kind,
            "amount": amount,
            "status": "created",
        }

    def count(self) -> int:
        """Сколько автодействий создано."""
        return len(self.actions)

File: assets/rule-templates/consent-before-auto-action/java/src/main/java/Fakes.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Фейки реестра согласий, незавершённых операций и счётчика автодействий.
 *
 * <p>Фейки считают то, что проверяет инвариант: есть ли действующая запись
 * о согласии ({@link ConsentRegistry#isActive}), открыта ли операция клиента
 * ({@link FakeOperations#hasOpen}) и сколько автодействий создано
 * ({@link FakeActionLog#count()}). Ни сети, ни сна.
 */
public final class Fakes {

    private Fakes() {
    }

    /** Созданное автодействие. */
    public record Action(String clientId, String kind, long amount, String status) {
    }

    /** Запись о согласии. */
    public record Consent(String consentId, String status) {

        /** Действует ли согласие. */
        public boolean active() {
            return "active".equals(status);
        }
    }

    /**
     * Реестр согласий: записи о согласии клиента на автодействие.
     *
     * <p>Запись хранит статус, а не только факт наличия: отозванное согласие
     * остаётся в реестре, но действующим не считается.
     */
    public static final class ConsentRegistry {

        private final Map<String, Consent> records = new HashMap<>();

        private static String key(String clientId, String scope) {
            return clientId + "|" + scope;
        }

        /** Записать согласие клиента в объёме {@code scope}. */
        public void grant(String clientId, String scope, String consentId) {
            records.put(key(clientId, scope), new Consent(consentId, "active"));
        }

        /** Отозвать согласие; запись остаётся со статусом {@code revoked}. */
        public void revoke(String clientId, String scope) {
            Consent record = records.get(key(clientId, scope));
            if (record != null) {
                records.put(key(clientId, scope), new Consent(record.consentId(), "revoked"));
            }
        }

        /** Есть ли действующая запись о согласии. */
        public boolean isActive(String clientId, String scope) {
            Consent record = records.get(key(clientId, scope));
            return record != null && record.active();
        }

        /** Запись о согласии или {@code null}. */
        public Consent recordFor(String clientId, String scope) {
            return records.get(key(clientId, scope));
        }
    }

    /**
     * Незавершённые операции клиента.
     *
     * <p>Пока операция клиента открыта, новое автодействие по нему не стартует.
     */
    public static final class FakeOperations {

        private final Set<String> open = new HashSet<>();

        /** Открыть операцию клиента. */
        public void open(String clientId, String operationId) {
            open.add(clientId);
        }

        /** Закрыть операцию клиента. */
        public void close(String clientId) {
            open.remove(clientId);
        }

        /** Открыта ли у клиента незавершённая операция. */
        public boolean hasOpen(String clientId) {
            return open.contains(clientId);
        }
    }

    /** Счётчик созданных автодействий. */
    public static final class FakeActionLog {

        private final List<Action> actions = new ArrayList<>();

        /** Создать автодействие и записать его в счётчик. */
        public Action create(String clientId, String kind, long amount) {
            Action action = new Action(clientId, kind, amount, "created");
            actions.add(action);
            return action;
        }

        /** Сколько автодействий создано. */
        public int count() {
            return actions.size();
        }

        /** Созданные автодействия, в порядке создания. */
        public List<Action> actions() {
            return List.copyOf(actions);
        }
    }
}
